### PR TITLE
SCP - 914

### PR DESCRIPTION
**Pull request by outbools**
Work in Progress.
http://www.scp-wiki.net/scp-914
Missing:
- [x] Refactor to component datums
- [x] Actually compile this
- [ ] Sprites with sane size
- [ ] Animations
- [ ] Sounds
- [x] Any transformation recipes

Put things into INTAKE booth, click on panel in the middle - take resulting product from OUTPUT booth

### RECIPES:
1. find an item you want to use.
2. put following lists to item's definition:
```
  rough = list(<list of items for that mode>)
  (required) coarse = list(...)
  (required) oneone = list(...)
  (required) fine = list(...)
  veryfine = list(...)
```
You may use only one list, but then selected item would be transformed only on mode where corresponding list is defined.
By default Rough/Very Fine will process item on Coarse/Fine twice, define rough/veryfine lists to specify any desired transformation
### SPECIAL EFFECTS:
Proc `/atom/scp914_act(var/mode)` ~can be overriden~CAN'T BE OVERRIDEN, override `/atom/scp914_act_on_type(var/atom/output)` to define any additional effects by your desire.
**Example**: Transfering mind from intake mob to output mob:
```
/mob/scp914_act_on_type(var/atom/output)
	if(ismob(output))
		var/mob/M = new output(loc)
		qdel(output)
		M.key = key
		return M
	return src
```

It wouldn't look THAT big of course
image
image
and the switch arrow is already fixed

### CODE REVIEW

**Comment by PigeonLord** [comment]
That's a biiiiiiig fucking SCP

**Comment by outbools** [comment]
Okay looks like i need to refactor this thing to fit component design

**Comment by JakeDaBoss** [comment]
God damnit, this'll be hell to configure for most items in SCP13. *grumble grumble*

**Comment by outbools** [comment]
OP updated with tutorials for new procs

**Comment by MrFroztee** [comment]
I'll keep up some pressure for an animated version of 914, sound should be easy

File: code/modules/SCP/SCPs/SCP-914.dm [hunk 1]
```diff
@@ -0,0 +1,251 @@
...
+
+/datum/component/scp/SCP_914
+	signal_procs = list(COMSIG_PARENT_ATTACKBY = .proc/transform)
+	var/clockwork = FALSE
+
+/datum/component/scp/SCP_914/proc/transform(obj/item/W, mob/user, var/click_params)
+  if(!W || clockwork)
+    return
+  var/atom/movable/product = W.scp914_act(rand(CLOCKMODE_ROUGH,CLOCKMODE_VERYFINE))
+  to_chat(user, "Your [W] turned into [product]!")
```
Comment (PJB3005) [inline code/modules/SCP/SCPs/SCP-914.dm]: Wouldn't it be much more fun to have to figure this out on your own?

File: code/modules/SCP/SCPs/SCP-914.dm [hunk 2]
```diff
@@ -0,0 +1,251 @@
...
+  classification = SAFE
+  component = /datum/component/scp/SCP_914
+
+/datum/component/scp/SCP_914
+	signal_procs = list(COMSIG_PARENT_ATTACKBY = .proc/transform)
+	var/clockwork = FALSE
+
+/datum/component/scp/SCP_914/proc/transform(obj/item/W, mob/user, var/click_params)
+  if(!W || clockwork)
+    return
```
Comment (PJB3005) [inline code/modules/SCP/SCPs/SCP-914.dm]: Is GitHub failing to display this or is this file's indentation messed up completely?

File: code/game/atoms.dm [hunk 3]
```diff
@@ -293,6 +301,69 @@ its easier to just keep the beam vertical.
...
+				if(choice)
+					var/atom/atombychoice = new choice(loc)
+					if(atombychoice)
+						return scp914_act_on_type(atombychoice)
+			if(3)
+				var/list/choices = subtypesof(parent_type)
+				if(choices && choices.len > 0)
+					for(var/atom/typo in choices)
+						var/atom/choico = new typo
+						if("[choico.parent_type]" != "[parent_type]")
```
Comment (PJB3005) [inline code/game/atoms.dm]: DM allows you to compare type paths. Sure you need to convert to string here?

File: code/game/atoms.dm [hunk 4]
```diff
@@ -293,6 +301,69 @@ its easier to just keep the beam vertical.
 /atom/proc/fire_act()
 	return
 
+/atom/proc/scp914_act(var/mode) //DON'T OVERRIDE, override scp914_act_on_type instead
+	var/list/chosen_mode = modes_914[mode]
+	if(chosen_mode && chosen_mode.len > 0)
```
Comment (PJB3005) [inline code/game/atoms.dm]: `length()` returns 0 for null so `length(chosen_mode) > 0` is perfect here.

File: code/game/atoms.dm [hunk 5]
```diff
@@ -293,6 +301,69 @@ its easier to just keep the beam vertical.
...
+								return scp914_act_on_type(atombychoice)
+			if(5) //TODO: Very Fine should give one SCP component to an output object
+				var/choice = pick(subtypesof(src))
+				if(choice)
+					var/atom/atombychoice = new choice(loc)
+					if(atombychoice)
+						return scp914_act_on_type(atombychoice)
+	return
+
+/atom/proc/scp914_act_on_type(var/atom/output) //to get around various type requirements, input/intake is src so don't worry
```
Comment (PJB3005) [inline code/game/atoms.dm]: You might want to provide the mode used to select this type *and* have another API to programmatically select the type based on mode.

File: code/game/atoms.dm [hunk 6]
```diff
@@ -293,6 +301,69 @@ its easier to just keep the beam vertical.
...
+	var/list/chosen_mode = modes_914[mode]
+	if(chosen_mode && chosen_mode.len > 0)
+		var/chosen = pick(chosen_mode)
+		var/atom/chosen_obj = new chosen(loc)
+		if(chosen_obj)
+			return scp914_act_on_type(chosen_obj)
+	else
+		switch(mode)
+			if(1) //TODO: 1 and 2 should be refactored to fit MATERIAL datums once they are ready
+				var/atom/choice = new parent_type
```
[the remark this answers is not in the document]
Comment (outbools) [reply code/game/atoms.dm]: it's fun

File: code/modules/SCP/SCPs/SCP-914.dm (hunk 1 above)
Comment (outbools) [reply code/modules/SCP/SCPs/SCP-914.dm]: they told me to shit some component to have fun with and i got zero idea how to put an actual component into actual 914 to work with.

oh you mean chat message... probably more fun, yes

File: code/game/atoms.dm (hunk 3 above)
Comment (outbools) [reply code/game/atoms.dm]: They do accepts children, something i want to avoid.

File: code/game/atoms.dm (hunk 4 above)
Comment (outbools) [reply code/game/atoms.dm]: got it

File: code/game/atoms.dm (hunk 5 above)
Comment (outbools) [reply code/game/atoms.dm]: oh right mode is important here

File: code/game/atoms.dm [hunk 7]
```diff
@@ -12,6 +12,14 @@
 	var/fluorescent // Shows up under a UV light.
 	var/datum/scp/SCP //For SCP's
 
+	var/list/rough
```
[the remark this answers is not in the document]
Comment (outbools) [reply code/game/atoms.dm]: and i wondering what the fuck is going on in game

File: code/game/atoms.dm [hunk 8]
```diff
@@ -293,6 +301,69 @@ its easier to just keep the beam vertical.
 /atom/proc/fire_act()
 	return
 
+/atom/proc/scp914_act(var/mode) //DON'T OVERRIDE, override scp914_act_on_type instead
```
[the remark this answers is not in the document]
Comment (outbools) [reply code/game/atoms.dm]: got it

File: code/game/atoms.dm [hunk 9]
```diff
@@ -293,6 +301,69 @@ its easier to just keep the beam vertical.
...
+	var/list/chosen_mode = modes_914[mode]
+	if(chosen_mode && chosen_mode.len > 0)
+		var/chosen = pick(chosen_mode)
+		var/atom/chosen_obj = new chosen(loc)
+		if(chosen_obj)
+			return scp914_act_on_type(chosen_obj)
+	else
+		switch(mode)
+			if(1) //TODO: 1 and 2 should be refactored to fit MATERIAL datums once they are ready
```
[the remark this answers is not in the document]
Comment (outbools) [reply code/game/atoms.dm]: got it

File: code/game/atoms.dm [hunk 10]
```diff
@@ -45,7 +53,7 @@
 
 	if(opacity)
 		updateVisibility(src)
-
+	modes_914 = list(rough, coarse, oneone, fine, veryfine)
```
[the remark this answers is not in the document]
Comment (outbools) [reply code/game/atoms.dm]: yeah you right
anyway i ended up with switch

File: code/game/atoms.dm (hunk 3 above)
Comment (JakeDaBoss) [reply code/game/atoms.dm]: Is that a hint at our staff? I tried working things out with you, if you had returned my messages.

**Comment by MrFroztee** [comment]
Closed due to contributor not actually caring about this repo anymore.

**Pull request closed without merging**